device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -606,17 +657,112 @@ public void registerConnectionStateCallback(IotHubConnectionStateCallback callba
      * @param callback the callback to be called. Can be null if callbackContext is not null
      * @param callbackContext a context to be passed to the callback. Can be {@code null}.
      */
-    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext, String deviceId)
     {
         if (callbackContext != null && callback == null)
         {
             //Codes_SRS_IOTHUBTRANSPORT_34_051: [If the provided callback is null but the context is not, this function shall throw an IllegalArgumentException.]
             throw new IllegalArgumentException("Callback cannot be null if callback context is null");
         }
 
-        //Codes_SRS_IOTHUBTRANSPORT_34_052: [This function shall save the provided callback and context.]
-        this.connectionStatusChangeCallback = callback;
-        this.connectionStatusChangeCallbackContext = callbackContext;
+        if (callback == null)
+        {
+            this.connectionStatusChangeCallbacks.remove(deviceId);
+            this.connectionStatusChangeCallbackContexts.remove(deviceId);
+        }
+        else
+        {
+            this.connectionStatusChangeCallbacks.put(deviceId, callback);
+
+            if (callbackContext != null)
+            {
+                // ConcurrentHashMaps don't support null values. If user provides null context,
+                // then calls to connectionStatusChangeCallbackContexts.get(...) will return null which lets this layer still work as expected.
+                this.connectionStatusChangeCallbackContexts.put(deviceId, callbackContext);
+            }
+
+        }
+    }
+
+    public void registerMultiplexingConnectionStateCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
+    {
+        if (callback == null && callbackContext != null)
+        {
+            throw new IllegalArgumentException("Cannot have a null callback and a non-null context associated with it");
+        }
+
+        this.multiplexingStateCallback = callback;
+        this.multiplexingStateCallbackContext = callbackContext;
+    }
+
+    public void registerMultiplexedDeviceClient(List<DeviceClientConfig> configs) throws InterruptedException {
+        if (getProtocol() != IotHubClientProtocol.AMQPS && getProtocol() != IotHubClientProtocol.AMQPS_WS)
```

Comment:
q - would the user ever create an instance of `IotHubTransport` directly? The reason I ask is, we already have these checks in the higher layers, so I was wondering if this check was even required.

Reply:
Customer's wouldn't create this IotHubTransport object directly, no. These checks aren't something that a customer would hit, but it may be hit by a developer who does something wrong (like tries to multiplex over MQTT or something.)

Reply:
I can remove it if you'd prefer, but this error message is more helpful to a developer on our team than a class cast failure exception that they'd get when they try to open later.

Reply:
No, you are right, it'll help against accidental misuse within the library.